[ResponseOps][Reporting][9.1 & Serverless] Scheduled Exports External Docs (#2118)
elastic/docs-content: explore-analyze/report-and-share/automating-report-generation.md

# Automatically generate reports [automating-report-generation]

> To automatically generate PDF and CSV reports, generate a POST URL, then submit an HTTP `POST` request using {{watcher}} or a script. In 9.1.0 and Serverless, you can use {{kib}} to generate reports on a recurring schedule and share them with a list of emails that you specify.

Reviewer: ```suggestion
To automatically generate PDF and CSV reports, generate a POST URL, then submit an HTTP `POST` request using {{watcher}} or a script. In {{stack}} 9.1 and Serverless, you can use {{kib}} to generate reports on a recurring schedule and share them with a list of emails that you specify.
```

This is how it's called out in the previous doc. Not sure if you want to keep that formatting here as well.

## Create a POST URL [create-a-post-url]

Create the POST URL that triggers a report to generate PDF and CSV reports.

### PDF reports

To create the POST URL for PDF reports:

1. Go to **Dashboards**, **Visualize Library**, or **Canvas**.
2. Open the dashboard, visualization, or **Canvas** workpad you want to view as a report. From the toolbar, do one of the following:

>     * {applies_to}`stack: ga 9.0` If you are using **Dashboard** or **Visualize Library**, click **Share > Export**, select the PDF option then click **Copy POST URL**.

Reviewer: ```suggestion
    * {applies_to}`stack: ga 9.0` If you are using **Dashboard** or **Visualize Library**, click **Share > Export**, select the PDF option, then click **Copy POST URL**.
```

    * {applies_to}`stack: ga 9.0` If you are using **Canvas**, click **Share > PDF Reports**, then click **Advanced options > Copy POST URL**.

>     * {applies_to}`stack: ga 9.0` Click the **Export** icon, then **PDF**. In the export flyout, copy the POST URL.

Reviewer 2: the export icon is ga 9.1 I believe

Reviewer 2: * {applies_to}`stack: ga 9.1` Click the **Export** icon, then **PDF**. In the export flyout, copy the POST URL.

This is what I meant by the export icon is ga 9.1

Author: ```suggestion
    * {applies_to}`stack: ga 9.1` Click the **Export** icon, then **PDF**. In the export flyout, copy the POST URL.
```

:::{important}
In earlier {{kib}} versions, you could use the `&sync` parameter to append to report URLs that held the request open until the document was fully generated. The `&sync` parameter is now unsupported. If you use the `&sync` parameter in Watcher, you must update the parameter.
:::

## Schedule and share reports [schedule-report-generation]

```{applies_to}

> stack: preview 9.1

Reviewer: Should there be a serverless `applies_to` tag? I noticed Serverless is mentioned in the requirements.

Reviewer 3: It's not mandatory as per the guidelines since it's implied from the page's frontmatter that all of the content applies to serverless unless specifically called out, but up to you @nastasha-solomon. I think it doesn't hurt to also add it just to avoid users think this is for stack 9.1 ONLY.

Author: Thanks, both! I figured that adding Serverless would be redundant since it's already mentioned in the front matter. But I can also see how users might miss that if they land directly on this section, or are still learning how the `applies to` tags work.

```suggestion
stack: preview 9.1
serverless: preview
```

Reviewer 3: Indeed if the feature described specifically in this section is in preview state in serverless, then it needs to be specifically called out 👍

Author: Ah, good point. Thanks for the reminder.

```

Save time by setting up a recurring task that automatically generates reports and shares them on a schedule that you choose. 

### Requirements [scheduled-reports-reqs]

> * To use the scheduled reports feature, your role needs [access to reporting](../../deploy-manage/kibana-reporting-configuration.md#grant-user-access)

Reviewer: ```suggestion
* To use the scheduled reports feature, your role needs [access to reporting](../../deploy-manage/kibana-reporting-configuration.md#grant-user-access).
```

* (Optional) To view and manage other users’ reports and schedules, your role needs `All` privileges for the **Manage Scheduled Reports** feature. You can set this by configuring your role's {{kib}} privileges.
* Sharing reports outside of {{kib}} requires a default preconfigured email connector.

   * **{{ech}} or {{serverless-short}} users**: You do not need to set up a default preconfigured email connector. Kibana provides you with a built-in preconfigured email connector that uses the SMTP protocol to send emails. To view it, go to the **Connectors** page and find the Elastic-Cloud-SMTP connector.
   * **Self-managed users**: You must set up a default preconfigured email connector to send notifications outside of {{kib}}. To do this:
     

>      1. Open your `kibana.yml` file. and add a new section for the `xpack.actions.preconfigured` setting. This setting specifies configuration details that are specific to the type of preconfigured connector you're defining. 

Reviewer: ```suggestion
     1. Open your `kibana.yml` file and add a new section for the `xpack.actions.preconfigured` setting. This setting specifies configuration details that are specific to the type of preconfigured connector you're defining. 
```

>      2. Under the `xpack.actions.preconfigured setting`, define the email connector for example: 

Reviewer: ```suggestion
     2. Under the `xpack.actions.preconfigured setting`, define the email connector. For example: 
```

:::{important}
In earlier {{kib}} versions, you could use the `&sync` parameter to append to report URLs that held the request open until the document was fully generated. The `&sync` parameter is now unsupported. If you use the `&sync` parameter in Watcher, you must update the parameter.
:::

## Schedule and share reports [schedule-report-generation]

```{applies_to}
stack: preview 9.1
serverless: preview
```

Save time by setting up a recurring task that automatically generates reports and shares them on a schedule that you choose. 

### Requirements [scheduled-reports-reqs]

* To use the scheduled reports feature, your role needs [access to reporting](../../deploy-manage/kibana-reporting-configuration.md#grant-user-access).

> * (Optional) To view and manage other users’ reports and schedules, your role needs `All` privileges for the **Manage Scheduled Reports** feature. You can set this by configuring your role's {{kib}} privileges.

Reviewer 2: To view , manage other user's reports and schedules , and to share report with others via email, your role needs ...etc

   * **Self-managed users**: You must set up a default preconfigured email connector to send notifications outside of {{kib}}. To do this:
     
     1. Open your `kibana.yml` file and add a new section for the `xpack.actions.preconfigured` setting. This setting specifies configuration details that are specific to the type of preconfigured connector you're defining. 
     2. Under the `xpack.actions.preconfigured setting`, define the email connector. For example: 

        ````
        xpack.actions.preconfigured:
          my-email:
          name: preconfigured-email-connector-type
          actionTypeId: .email
          config:
            service: other
            from: testsender@test.com
            host: validhostname
            port: 8080
            secure: false
            hasAuth: true
          secrets:
            user: testuser
            password: passwordkeystorevalue
         `````

         :::{tip} 

>          Refer to [Email connectors](kibana://reference/connectors-kibana/pre-configured-connectors.md#preconfigured-email-configuration) to learn about requirements for different email services and providers.

Author: Added in [c7e576f](https://github.com/elastic/docs-content/pull/2118/commits/c7e576f9c7b37c8cbe8cbfa140ca33ebed3cc42d).

         :::

### Create a schedule [create-scheduled-report]

1. Open the saved Discover session, dashboard, or visualization you want to share. 
2. Click the **Export** icon, then **Schedule export**.
3. Enter the requested details, and (optional) enable **Print format** to generate the report in a printer-friendly format.
4. Set up a schedule for generating the report.

    * **Date**: Choose when to start generating reports.
    * **Timezone**: Specify a timezone for the schedule.
    * **Repeat**: Choose how often you want to generate reports.  

> 5. (Optional) To share generated reports outside of Kibana, enable **Send by email** and enter a list of email addresses. Recipients will receive emails with the generated reports attached and on the schedule that you specified.

Reviewer 4: We should mention that users without `Manage Scheduled Reporting` feature privilege will only be able to send the report to themselves.

Author: Added in [c7e576f](https://github.com/elastic/docs-content/pull/2118/commits/c7e576f9c7b37c8cbe8cbfa140ca33ebed3cc42d).

6. Click **Schedule exports** to save the schedule. 

A message appears, indicating that the schedule is available on the **Reporting** page. From the **Reporting** page, click on the **Schedules** tab to view details for the newly-created schedule. 

::::{important} 

> Note that you cannot edit or delete a schedule after you create it. To stop the schedule from running, you must disable it. Disabling a schedule permanently stops it from running. To restart it, you must create a new schedule. 

Reviewer 4: Should we add a section on limitations? For example, we only allow attachments up to a certain size. Also, because we run reporting tasks one at a time, if they schedule too many reports for the same time, they may receive their reports on a delay. In general, if their cluster is busy with a lot of tasks, their reports may be delayed.

Author: Added in [006df70](https://github.com/elastic/docs-content/pull/2118/commits/006df70eab1ba3e54d62a1d1a40b34b9d4a2c7f9).

:::{important}
In earlier {{kib}} versions, you could use the `&sync` parameter to append to report URLs that held the request open until the document was fully generated. The `&sync` parameter is now unsupported. If you use the `&sync` parameter in Watcher, you must update the parameter.
:::

## Schedule and share reports [schedule-report-generation]

```{applies_to}
stack: preview 9.1
serverless: preview
```

Save time by setting up a recurring task that automatically generates reports and shares them on a schedule that you choose. 

### Requirements [scheduled-reports-reqs]

* To use the scheduled reports feature, your role needs [access to reporting](../../deploy-manage/kibana-reporting-configuration.md#grant-user-access).

> * (Optional) To view and manage other users’ reports and schedules, your role needs `All` privileges for the **Manage Scheduled Reports** feature. You can set this by configuring your role's {{kib}} privileges. If your role doesn't have the **Manage Scheduled Reporting** feature privilege, you can only share reports with yourself. 

Author: ```suggestion
* (Optional) To view and manage other user's reports and schedules, and to share reports with others using email, your role needs `All` privileges for the **Manage Scheduled Reports** feature. You can set this by configuring your role's {{kib}} privileges. If your role doesn't have the **Manage Scheduled Reporting** feature privilege, you can only share reports with yourself. 
```

:::{important}
In earlier {{kib}} versions, you could use the `&sync` parameter to append to report URLs that held the request open until the document was fully generated. The `&sync` parameter is now unsupported. If you use the `&sync` parameter in Watcher, you must update the parameter.
:::

> ## Schedule and share reports [schedule-report-generation]

Reviewer 5: Wondering if it would be worth mentioning which report types are compatible with scheduling 🤔 There's only one non-compatible type which is CSV in Lens

Author: This would be a good detail to include in the limitations section. I'll add it - thanks!

     The following example shows a modified `kibana.yml` file with a preconfigured email connector that's set as the default connector for email notifications:

      ```
        xpack.actions.preconfigured:
          my-email:
            name: preconfigured-email-connector-type
            actionTypeId: .email
            config:
              service: other
              from: testsender@test.com
              host: validhostname
              port: 8080
              secure: false
              hasAuth: true
            secrets:
              user: testuser
              password: passwordkeystorevalue

        notifications.connectors.default.email: my-email
     ```

### Create a schedule [create-scheduled-report]

1. Open the saved Discover session, dashboard, or visualization you want to share. 
2. Click the **Export** icon, then **Schedule export**.

> 3. Enter the requested details, and (optional) enable **Print format** to generate the report in a printer-friendly format.

Author: The **Print format** toggle disappears if you select the PNG option so I don't think it's necessary to call out in the docs. : )

4. Set up a schedule for generating the report.

    * **Date**: Choose when to start generating reports.
    * **Timezone**: Specify a timezone for the schedule.
    * **Repeat**: Choose how often you want to generate reports.  

> 5. (Optional) To share generated reports outside of Kibana, enable **Send by email** and enter a list of email addresses. Recipients will receive emails with the generated reports attached and on the schedule that you specified.

Reviewer 5: Do we want to specify anything about the emails validation (allowlist) here?

* (Optional) To view and manage other users’ reports and schedules, your role needs `All` privileges for the **Manage Scheduled Reports** feature. You can set this by configuring your role's {{kib}} privileges. If your role doesn't have the **Manage Scheduled Reporting** feature privilege, you can only share reports with yourself. 
* Sharing reports outside of {{kib}} requires a default preconfigured email connector.

   * **{{ech}} or {{serverless-short}} users**: You do not need to set up a default preconfigured email connector. Kibana provides you with a built-in preconfigured email connector that uses the SMTP protocol to send emails. To view it, go to the **Connectors** page and find the Elastic-Cloud-SMTP connector.
   * **Self-managed users**: You must set up a default preconfigured email connector to send notifications outside of {{kib}}. To do this:
     
     1. Open your `kibana.yml` file.
     2. Add the `xpack.actions.preconfigured` {{kib}} setting. This setting specifies configuration details for the preconfigured connector that you're defining. 

>      2. Under the `xpack.actions.preconfigured` setting, define the email connector. Refer to [Email connectors](kibana://reference/connectors-kibana/pre-configured-connectors.md#preconfigured-email-configuration) to learn about requirements for different email services and providers.

Reviewer: There are two step 2s, but they seem to render correctly. I knew you could just use 1. for every step, but didn't know it would render correctly in this instance. TIL!

         You must define preconfigured email connector details in the `kibana.yml` file. You cannot create a preconfigured email connector from the {{kib}} UI. 
         :::

     3. Add the `notifications.connectors.default.email` {{kib}} setting, and provide the name of your email connector. The `notifications.connectors.default.email` setting specifies the default email connector to use when sending notifications. This is especially useful if you have multiple email connectors and want to set a default one. 

     The following example shows a modified `kibana.yml` file with a preconfigured email connector that's set as the default connector for email notifications:

      ```
        xpack.actions.preconfigured:
          my-email:
            name: preconfigured-email-connector-type
            actionTypeId: .email
            config:
              service: other
              from: testsender@test.com
              host: validhostname
              port: 8080
              secure: false
              hasAuth: true
            secrets:
              user: testuser
              password: passwordkeystorevalue

        notifications.connectors.default.email: my-email
     ```

> * (Optional) To control who can receive email notifications from {{kib}}, add the [`xpack.actions.email.domain_allowlist` setting](kibana://settings/alert-action-settings.asciidoc) to your `kibana.yml` file. To learn more about configuring this setting, refer to [Notifications domain allowlist](../alerts-cases/alerts/notifications-domain-allowlist.md).

Author: ```suggestion
* (Optional) To control who can receive email notifications from {{kib}}, add the [`xpack.actions.email.domain_allowlist` setting](kibana://reference/configuration-reference/alerting-settings.md) to your `kibana.yml` file. To learn more about configuring this setting, refer to [Notifications domain allowlist](../alerts-cases/alerts/notifications-domain-allowlist.md).
```

> * (Optional) To control who can receive email notifications from {{kib}}, add the [`xpack.actions.email.domain_allowlist` setting](kibana://reference/configuration-reference.alerting-settings.md) to your `kibana.yml` file. To learn more about configuring this setting, refer to [Notifications domain allowlist](../alerts-cases/alerts/notifications-domain-allowlist.md).

Author: ```suggestion
* (Optional) To control who can receive email notifications from {{kib}}, add the [`xpack.actions.email.domain_allowlist` setting](kibana://settings/alert-action-settings.asciidoc) to your `kibana.yml` file. To learn more about configuring this setting, refer to [Notifications domain allowlist](../alerts-cases/alerts/notifications-domain-allowlist.md).
```
